src/OrchardCore/OrchardCore.Environment.Shell/ShellHost.cs

```diff
@@ -72,13 +72,23 @@ public void Initialize()
 
         public ShellContext GetOrCreateShellContext(ShellSettings settings)
         {
-            return _shellContexts.GetOrAdd(settings.Name, tenant =>
+            var shell = _shellContexts.GetOrAdd(settings.Name, tenant =>
             {
                 var shellContext = CreateShellContextAsync(settings).Result;
                 RegisterShell(shellContext);
 
                 return shellContext;
             });
+
+            if (shell.Released && shell.ActiveScopes == 0)
+            {
+                if (_shellContexts.TryRemove(settings.Name, out var context))
+                {
+                    return GetOrCreateShellContext(settings);
```

Comment:
We should call `_shellContexts.TryRemove` but always do `return GetOrCreateShellContext(settings)` even if `TryRemove` failed, meaning it was already removed. We still need to return a valid ShellContext.